src/Workspaces/Remote/Core/SolutionAssetStorage.cs

```diff
@@ -0,0 +1,188 @@
+            if (remotableData != null)
+            {
+                return remotableData;
+            }
+
+            // if it reached here, it means things get cancelled. due to involving 2 processes,
+            // current design can make slightly staled requests to running even when things cancelled.
+            // if it is other case, remote host side will throw and close connection which will cause
+            // vs to crash.
+            // this should be changed once I address this design issue
+            cancellationToken.ThrowIfCancellationRequested();
+
+            return null;
+        }
+
+        /// <summary>
+        /// Retrieve assets of specified <paramref name="checksums"/> available within <paramref name="scopeId"/> scope from the storage.
+        /// </summary>
+        public async ValueTask<IReadOnlyDictionary<Checksum, SolutionAsset>> GetAssetsAsync(int scopeId, IEnumerable<Checksum> checksums, CancellationToken cancellationToken)
+        {
+            using var checksumsToFind = Creator.CreateChecksumSet(checksums);
+
+            var numberOfChecksumsToSearch = checksumsToFind.Object.Count;
+            var result = new Dictionary<Checksum, SolutionAsset>(numberOfChecksumsToSearch);
+
+            if (checksumsToFind.Object.Remove(Checksum.Null))
+            {
+                result[Checksum.Null] = SolutionAsset.Null;
+            }
+
+            await FindAssetsAsync(_solutionStates[scopeId], checksumsToFind.Object, result, cancellationToken).ConfigureAwait(false);
+            if (result.Count == numberOfChecksumsToSearch)
+            {
+                // no checksum left to find
+                Debug.Assert(checksumsToFind.Object.Count == 0);
+                return result;
+            }
+
+            // if it reached here, it means things get cancelled. due to involving 2 processes,
+            // current design can make slightly staled requests to running even when things cancelled.
+            // if it is other case, remote host side will throw and close connection which will cause
+            // vs to crash.
+            // this should be changed once I address this design issue
+            cancellationToken.ThrowIfCancellationRequested();
+
+            return result;
+        }
+
+        /// <summary>
+        /// Find an asset of the specified <paramref name="checksum"/> within <paramref name="solutionState"/>.
+        /// </summary>
+        private static async ValueTask<SolutionAsset?> FindAssetAsync(SolutionState solutionState, Checksum checksum, CancellationToken cancellationToken)
+        {
+            using var checksumPool = Creator.CreateChecksumSet(SpecializedCollections.SingletonEnumerable(checksum));
+            using var resultPool = Creator.CreateResultSet();
+
+            await FindAssetsAsync(solutionState, checksumPool.Object, resultPool.Object, cancellationToken).ConfigureAwait(false);
+
+            if (resultPool.Object.Count == 1)
```

Comment:
1. can we assert that count should be 0 or 1.
2. what does it mean to get 0 back?  does that indicate a bug of some sort?  i.e .someone asking for something based on checksum, but we don't have that value?  i.e. is this masking an issue?  should we assert?

Reply:
https://github.com/dotnet/roslyn/issues/47013